keel-orca/src/main/kotlin/com/netflix/spinnaker/keel/orca/OrcaTaskMonitorAgent.kt

```diff
@@ -72,37 +74,57 @@ class OrcaTaskMonitorAgent(
             }.await()
         }
         .filterValues { it != null && it.status.isComplete() }
-        .map { (resourceId, taskDetails) ->
-          if (taskDetails != null) {
+        .map { (taskRecord, executionDetails) ->
+          if (executionDetails != null) {
             // only resource events are currently supported
-            if (resourceId.startsWith(RESOURCE.toString())) {
-              val id = resourceId.substringAfter(":")
+            if (taskRecord.subject.startsWith(RESOURCE.toString())) {
+              val resourceId = taskRecord.subject.substringAfter(":")
               try {
-                when (taskDetails.status.isSuccess()) {
-                  true -> publisher.publishEvent(
-                    ResourceTaskSucceeded(
-                      resourceRepository.get(id), listOf(Task(taskDetails.id, taskDetails.name)), clock
-                    )
-                  )
-                  false -> publisher.publishEvent(
-                    ResourceTaskFailed(
-                      resourceRepository.get(id),
-                      taskDetails.execution.stages.getFailureMessage()
-                        ?: "",
-                      listOf(Task(taskDetails.id, taskDetails.name)), clock
-                    )
-                  )
+                if (executionDetails.status.isSuccess()) {
+                  handleTaskSuccess(resourceId, executionDetails, taskRecord)
+                } else {
+                  handleTaskFailure(resourceId, executionDetails)
                 }
               } catch (e: NoSuchResourceId) {
                 log.warn("No resource found for id $resourceId")
               }
             }
-            taskTrackingRepository.delete(taskDetails.id)
+            taskTrackingRepository.delete(executionDetails.id)
           }
         }
     }
   }
 
+  private fun handleTaskSuccess(resourceId: String, executionDetails: ExecutionDetailResponse, taskRecord: TaskRecord) {
+    publisher.publishEvent(
+      ResourceTaskSucceeded(
+        resourceRepository.get(resourceId), listOf(Task(executionDetails.id, executionDetails.name)), clock
+      )
+    )
+
+    if (taskRecord.artifactVersionUnderDeployment != null) {
+      log.debug("Successful task ${taskRecord.id} deployed artifact version ${taskRecord.artifactVersionUnderDeployment} " +
+        "to resource ${resourceId}. Notifying artifact version deployed.")
+      publisher.publishEvent(
```

Comment:
This now handles publishing of the `ArtifactVersionDeployed` event from a single place independent of resource type.